src/battery-check/battery-check.c

```diff
@@ -0,0 +1,125 @@
+         * message than wait for plymouth */
+        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0) {
+                log_warning_errno(errno, "socket() failed: %m");
+                return;
+        }
+
+        if (connect(fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0) {
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Connection to plymouth failed: %m");
+                return;
+        }
+
+        if (write(fd, message, n + 1) != w)
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Failed to write to plymouth: %m");
+
+}
+
+static int parse_argv(int argc, char * argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",    no_argument, NULL, 'h'         },
+                { "version", no_argument, NULL, ARG_VERSION },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+
+                switch (c) {
+
+                case 'h':
+                        help();
+                        return 0;
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        if (optind < argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "%s takes no argument.",
+                                       program_invocation_short_name);
+        return 1;
+}
+
+static int run(int argc, char *argv[]) {
+        int r;
+
+        log_parse_environment();
+        log_open();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        r = battery_is_discharging_and_low();
+        if (r < 0) {
+                log_warning_errno(r, "Failed to check battery status, ignoring: %m");
+                return 0;
+        }
+        if (r > 0) {
+                _cleanup_close_ int fd = -EBADF;
+                log_emergency("Battery level critically low. Powering off in 10 seconds.");
+                fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
+                if (fd < 0)
+                        log_warning_errno(fd, "Failed to open console, ignoring: %m");
+                else
+                        dprintf(fd, "%s" ANSI_HIGHLIGHT_RED " Battery level critically low. Powering off in 10 seconds. " ANSI_NORMAL "%s\n", special_glyph(SPECIAL_GLYPH_LOW_BATTERY), special_glyph(SPECIAL_GLYPH_LOW_BATTERY));
+                battery_check_send_plymouth_message("Battery level critically low. Powering off in 10 seconds.");
```

Comment:
why no battery icon for plymouth? Aside from setting the message, you also want to change the mode to shutdown as well right?

Reply:
> why no battery icon for plymouth? Aside from setting the message, you also want to change the mode to shutdown as well right?

Do you mean in the socket communication to plymouth, we should set a shutdown mode? sorry if the questions sound naive.

Reply:
plymouth has multiple "modes".  During boot it's in "boot-up" mode where it may show an elaborate animation tied to boot progress. In "shutdown" mode it may just show a fast animation and go to a static frame and wait.

If we're shutting down because the battery is critical, it makes sense to tell plymouth to switch to "shutdown" mode so the themes can switch tacks.  To switch it to shutdown mode you need to send plymouth a message like
```
asprintf(&request, "C\002%cshutdown", sizeof ("shutdown"));
```

Reply:
@halfline , so all that we need do to switch plymouth to ```shutdown``` mode is to have the string "C\002shutdown" in our message?

could I just add make two ```write()``` calls?
one with the actual message, and one with the "shutdown" string?

Also I want to clarify that ```"C\002%c"``` for instance should be ```"C\002%d"``` since the corresponding argument is of type int? is that right sir?

I want to push an initial modification, that has the ```M\002``` 
/format and the battery emoji in our socket message,
kindly review sir.